core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1698,150 +1269,348 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
+      activeControllerId = -1
+      onControllerResignation()
+    }
+  }
+
 }
 
-object LogDirEventNotificationListener {
-  val version: Long = 1L
+class BrokerChangeHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChildChangeHandler {
+  override val path: String = BrokerIdsZNode.path
+
+  override def handleChildChange: Unit = eventManager.put(controller.BrokerChange)
 }
 
-class PartitionModificationsListener(controller: KafkaController, eventManager: ControllerEventManager, topic: String) extends IZkDataListener with Logging {
-  override def handleDataChange(dataPath: String, data: Any): Unit = {
-    eventManager.put(controller.PartitionModifications(topic))
-  }
+class TopicChangeHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChildChangeHandler {
+  override val path: String = TopicsZNode.path
 
-  override def handleDataDeleted(dataPath: String): Unit = {}
+  override def handleChildChange: Unit = eventManager.put(controller.TopicChange)
 }
 
-/**
-  * Delete topics includes the following operations -
-  * 1. Add the topic to be deleted to the delete topics cache, only if the topic exists
-  * 2. If there are topics to be deleted, it signals the delete topic thread
-  */
-class TopicDeletionListener(controller: KafkaController, eventManager: ControllerEventManager) extends IZkChildListener with Logging {
-  override def handleChildChange(parentPath: String, currentChilds: java.util.List[String]): Unit = {
-    import JavaConverters._
-    eventManager.put(controller.TopicDeletion(currentChilds.asScala.toSet))
-  }
+class LogDirEventNotificationHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChildChangeHandler {
+  override val path: String = LogDirEventNotificationZNode.path
+
+  override def handleChildChange: Unit = eventManager.put(controller.LogDirEventNotification)
 }
 
-/**
- * Starts the partition reassignment process unless -
- * 1. Partition previously existed
- * 2. New replicas are the same as existing replicas
- * 3. Any replica in the new set of replicas are dead
- * If any of the above conditions are satisfied, it logs an error and removes the partition from list of reassigned
- * partitions.
- */
-class PartitionReassignmentListener(controller: KafkaController, eventManager: ControllerEventManager) extends IZkDataListener with Logging {
-  override def handleDataChange(dataPath: String, data: Any): Unit = {
-    val partitionReassignment = ZkUtils.parsePartitionReassignmentData(data.toString)
-    eventManager.put(controller.PartitionReassignment(partitionReassignment))
-  }
+class PartitionModificationsHandler(controller: KafkaController, eventManager: ControllerEventManager, topic: String) extends ZNodeChangeHandler {
+  override val path: String = TopicZNode.path(topic)
 
-  override def handleDataDeleted(dataPath: String): Unit = {}
+  override def handleCreation: Unit = ???
+
```

Comment:
This will be called when a topic is deleted. So, we probably want to make this a no-op instead of throwing an exception.

Reply:
That's not quite right. TopicDeletionManager's `completeDeleteTopic` first unregisters the PartitionModificationsHandler before deleting the topic znode.

In addition to looking at the code, I did a topic deletion and didn't find any errors in any of the logs.

I think there are 3 options here:
1. change unimplemented methods to be no-ops on a case-by-case basis.
2. change all of the controller's unimplemented handle methods to be no-ops.
3. change the traits themselves to make these methods no-ops by default. You'd only override if you need to.

I am leaning towards option 3 since it'll make the code more concise but am okay with any option.

Reply:
Option 3 sounds reasonable. The thing with unregistering a watcher is that it only gets reflected during the next read. So, it may not happen immediately.

Reply:
Watches are tricky with zookeeper. The key point is that ZookeeperClient handler unregistration is not the same thing as watcher removal.

Prior to zookeeper 3.5, there actually [isn't a way to explicitly remove a session's watcher from the ensemble](https://issues.apache.org/jira/browse/ZOOKEEPER-442). Watchers can only get removed after they get triggered.

However, from the ZookeeperClient's perspective, handler unregistration is local and immediate. It's possible for the zookeeper ensemble to send you a notification for a watcher after you've unregistered that watcher's corresponding handler, but it won't have any effect. ZookeeperClient maintains local mappings from paths to registered handlers and updates these mappings immediately upon handler registration and unregistration. When ZookeeperClient receives a watcher notification, its ZookeeperClientWatcher first looks up a handler in its local mappings and only if one exists does it actually trigger the corresponding handle method.

Reply:
Ok, then, it's probably covered in this case.